docs: update data fetching for `$fetch` and refresh (#3914)
nuxt/framework: docs/content/3.docs/1.usage/1.data-fetching.md

</script>
```

> ## Refreshing Data

Reviewer: ```suggestion
## Refreshing Data

```

  </header>
</template>
```

> ## Changing URL Parameters with `asyncData`

Reviewer: We could make this section more general as "Refreshing Data". With the explanation of `refrash` callback and `watch` options.

As for the example, In order to properly trigger refrash on url and query changes, `useFetch` can be used:

```js
const { data } = await useFetch(() => `users?page=${page.value}&take=6`, { baseURL: config.API_BASE_URL }
);
```

We might add it as an example to `useFetch` section.

Author: @pi0 I moved this up under the `useFetch()` section.  Do you have an example of watching parameters? If so, I can inspect and explain it.

Sometimes throughout the course of your user's page visit, you may need to refresh the data loaded from the API. This can happen if the user chooses to paginate, filter results, search, etc. 

You can make use of the `refresh()` method returned from the `useAsyncData()` composable to refresh the data with different query parameters:

```vue
<script setup>
const page = ref(1);

> const { data:users, loading, refresh, error } = await useAsyncData( 'users', () => 

Reviewer: As for the example, In order to properly trigger refrash on url and query changes, `useFetch` can be used:

```js
const { data } = await useFetch(() => `users?page=${page.value}&take=6`, { baseURL: config.API_BASE_URL }
);
```

We might add it as an example to `useFetch` section.

Author: I got this updated in thanks!

  $fetch( 
    config.API_BASE_URL+`/users?page=${page.value}&take=6` 
  )
);

function previous(){
  page.value--;
  refresh();
}

function next() {
  page.value++;
  refresh();
}
</script>
```
The key to making this work is to call the `refresh()` method returned from the `useAsyncData()` composable when a query parameter has changed.

## Directly calling an API Endpoint

> There are instances where you may need to directly call the API. Nuxt 3 provides a globally available $fetch method that has the same API as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).

Reviewer: ```suggestion
There are instances where you may need to directly call the API. Next 3 provides globally available `$fetch` method using [unjs/ohmyfetch](https://github.com/unjs/ohmyfetch) as well as `fetch`
with the same API as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).
```

> It has a number of benefits, including:

Reviewer: ```suggestion
Using `$fetch` has a number of benefits, including:
```

  </header>
</template>
```

> ## Directly calling an API Endpoint

Reviewer: ```suggestion
## Directly calling an API Endpoint

```

  </header>
</template>
```

## Directly calling an API Endpoint

> There are instances where you may need to directly call the API. Next 3 provides globally available `$fetch` method using [unjs/ohmyfetch](https://github.com/unjs/ohmyfetch) as well as `fetch`
> with the same API as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).

Reviewer 2: ```suggestion
There are instances where you may need to directly call the API. Nuxt 3 provides a globally available `$fetch` method using [unjs/ohmyfetch](https://github.com/unjs/ohmyfetch) (in addition to `fetch`)
with the same API as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).
```

Using `$fetch` has a number of benefits, including:

It will handle 'smartly' making direct API calls if it's running on the server, or making a client-side call to your API if it's running on the client. (It can also handle calling third-party APIs.)

Plus, it comes with convenience features including automatically parsing responses and stringifying data.

You can also abstract your API requests into re-usable methods or composables that you can call from anywhere. Loading a paginated group of users would look something like:

```vue
<script setup>

>   const page = ref(1);
>   const users = ref({});
> 
>   async function loadUsers (page) {
>     const response = await $fetch(
>       '/api/v1/users',
>       {
>         method: 'GET',
>         params: {
>           page: page,
>           take: 6
>         }
>       }
>     );
> 
>     users.value = response;
>   }
> 
>   function next () {
>     page.value++;
>     loadUsers(page.value);
>   }
> 
>   function previous () {
>     page.value--;
>     loadUsers(page.value);
>   }
> 
>   loadUsers(page.value);

Reviewer 2: As there is only one instance of users I would suggest instead something like.

```suggestion
  const page = ref(1);
  const users = ref([]);

  watch(page, async (page) => {
    users.value = await $fetch(
      '/api/v1/users',
      {
        method: 'GET',
        params: {
          page,
          take: 6
        }
      }
    )
  }, { immediate: true })

  function next () {
    page.value++;
  }

  function previous () {
    page.value--;
  }
```

Reviewer 2: But I'd also point out that this isn't yet a reusable composable. Maybe it would be worth giving an example of that?

Reviewer: In regards of example above as daniel noted, it is not a reusable composable and directly using `useFetch` is better since it can automatically do the watching as well. 

> Also, In Nuxt 2, I made a plugin that wrapped our API endpoints, this.$api.{resource}.show(). This way I had a single source of truth whenever the API interface changed (i.e. /v1 to /v2). Would it be helpful to give an example about wrapping an entire API Resource (GET, POST, PATCH, DELETE endpoints) as a composable? 

This is a really nice pattern as well. I would like to have it in examples/ section at least! For this PR, i would suggest to keep changes minimal and mergable and iterate over with next ideas.
